drivers/source/BufferedSerial.cpp

```diff
@@ -0,0 +1,381 @@
+{
+    size_t data_written = 0;
+    const char *buf_ptr = static_cast<const char *>(buffer);
+
+    if (length == 0) {
+        return 0;
+    }
+
+    if (core_util_in_critical_section()) {
+        return write_unbuffered(buf_ptr, length);
+    }
+
+    api_lock();
+
+    // Unlike read, we should write the whole thing if blocking. POSIX only
+    // allows partial as a side-effect of signal handling; it normally tries to
+    // write everything if blocking. Without signals we can always write all.
+    while (data_written < length) {
+
+        if (_txbuf.full()) {
+            if (!_blocking) {
+                break;
+            }
+            do {
+                api_unlock();
+                // Should we have a proper wait?
+                thread_sleep_for(1);
+                api_lock();
+            } while (_txbuf.full());
+        }
+
+        while (data_written < length && !_txbuf.full()) {
+            _txbuf.push(*buf_ptr++);
+            data_written++;
+        }
+
+        core_util_critical_section_enter();
+        if (_tx_enabled && !_tx_irq_enabled) {
+            // only write to hardware in one place
+            BufferedSerial::tx_irq();
+            if (!_txbuf.empty()) {
+                enable_tx_irq();
+            }
+        }
+        core_util_critical_section_exit();
+    }
+
+    api_unlock();
+
+    return data_written != 0 ? (ssize_t) data_written : (ssize_t) - EAGAIN;
+}
+
+ssize_t BufferedSerial::read(void *buffer, size_t length)
+{
+    size_t data_read = 0;
+
+    char *ptr = static_cast<char *>(buffer);
+
+    if (length == 0) {
+        return 0;
+    }
+
+    api_lock();
+
+    while (_rxbuf.empty()) {
+        if (!_blocking) {
+            api_unlock();
+            return -EAGAIN;
+        }
+        api_unlock();
+        // Do we need a proper wait?
+        thread_sleep_for(1);
+        api_lock();
+    }
+
+    while (data_read < length && !_rxbuf.empty()) {
+        _rxbuf.pop(*ptr++);
+        data_read++;
+    }
+
+    core_util_critical_section_enter();
+    if (_rx_enabled && !_rx_irq_enabled) {
+        // only read from hardware in one place
+        BufferedSerial::rx_irq();
+        if (!_rxbuf.full()) {
+            enable_rx_irq();
+        }
+    }
+    core_util_critical_section_exit();
+
+    api_unlock();
+
+    return data_read;
+}
+
+bool BufferedSerial::hup() const
+{
+    return _dcd_irq && _dcd_irq->read() != 0;
+}
+
+void BufferedSerial::wake()
+{
+    if (_sigio_cb) {
+        _sigio_cb();
+    }
+}
+
+short BufferedSerial::poll(short events) const
+{
+
+    short revents = 0;
+    // Check the Circular Buffer if space available for writing out
+
+
+    if (!_rxbuf.empty()) {
+        revents |= POLLIN;
+    }
+
+    // POLLHUP and POLLOUT are mutually exclusive
+    if (hup()) {
+        revents |= POLLHUP;
+    } else if (!_txbuf.full()) {
+        revents |= POLLOUT;
+    }
+
+    // Should other event types be handled?
```

Comment:
I would remove the question/comment and add in the function header which events are currently handled.

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/22de8309872101e3ff50230a4c880f689b952936..a74ffacf81aae2a915d556be6e0eea8119d3d8d0).